### PR TITLE
REL-2780 reset corrupted keychains

### PR DESCRIPTION
**Pull request by tpolecat**
This PR adds logic to reset keychains that are unreadable or can't be deserialized following decryption.

### CODE REVIEW

**Comment by tpolecat** [comment]
Had I known better at the time I would have written this with `Free` … dealing with the transformer stack is kind of a pain.

File: bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  /** Test peers. */
+  val testPeers: Set[Peer] = Set(
+    new Peer("foo", 123, Site.GN), 
+    new Peer("bar", 456, null)
+  )
+
+  /** Construct a KeyChain with a copy of the resource at the provided path. */
+  def keychain(resourcePath: String): Action[KeyChain] = {
+    val path = Files.createTempFile("KeyChainSpec", ".ser")
+    val file = path.toFile; file.deleteOnExit()
```
Comment (cquiroz) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: weird to see a `;`

File: bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    val in   = getClass.getResourceAsStream(resourcePath)
+    Files.copy(in, path, StandardCopyOption.REPLACE_EXISTING)
+    in.close()
+    KeyChain(file, KeyChain.KeyFetcher.forTesting(null), testPeers.toList)
+  }
+
+  /** Test that the keychain at the given path, after preparation, has peers = testPeers. */
+  def testKeyChain(path: String, prep: KeyChain => Action[Unit] = _ => Action.unit) =
+    (keychain(path) >>! prep >>= (_.peers)).unsafeRunAndThrow must_== testPeers
```
Comment (cquiroz) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: What`s `>>!`

File: bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala [hunk 3]
```diff
@@ -411,6 +436,26 @@ object KeyChain {
       _ <- keychain.updateSubject // IMPORTANT, see note at declaration of subject above
     } yield keychain
 
+  def log(level: Level, s: String): Action[Unit] =
```
Comment (cquiroz) [inline bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala]: Why do these 2 log functions look so different? Are they both used?

**Comment by cquiroz** [comment]
Do you mean monad transformer stack? I can't comment if `Free` is easier to understand but whenever I see an `EitherT`, e.t. al my little brain just can't continue parsing

File: bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala (hunk 2 above)
Comment (tpolecat) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: `fa >>! f` is "bind for effects", so it's equivalent to `for { a <- fa; _ <- f(a) } yield a`

File: bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala (hunk 3 above)
Comment (tpolecat) [inline bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala]: Oops yeah I'm not using the passed `Level`. I'll fix that.

**Comment by tpolecat** [comment]
Yeah it started off a lot simpler and the `EitherT` was fine but as the logic got more complex it started to feel klunky. `Free` is a much nicer representation for these little languages.

**Comment by cquiroz** [comment]
I urge that before adopting something like `Free` we discuss it among ourselves to be aware of the cost for people not familiar with the concepts and practice. It maybe much clearer that the monad transformer though

File: bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala [hunk 4]
```diff
@@ -187,7 +190,15 @@ abstract class KeyChain(kCell: KeyChain.KCell, pCell: KeyChain.PCell, lCell: Key
     for {
       p <- pCell.get.map(_.getOrElse(Password.default))
       v <- kCell.get.map(_.get(p))
-      k <- v.fold(_ => Action.fail(KeyFailure.KeychainLocked), Action(_))
+      k <- v match {
+             case -\/(_: IOException) =>
+
+               // Kernel is corrupted; reset and retry. See comment in tryPass below.
+               KeyChain.warn("Discarding corrupted kernel.") *> reset(Some(p)) *> kernel
```
Comment (sraaphorst) [inline bundle/edu.gemini.util.security/src/main/scala/edu/gemini/util/security/auth/keychain/KeyChain.scala]: What does the `*>` operator do?

File: bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala (hunk 1 above)
Comment (sraaphorst) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: Agreed with @cquiroz . Could someone explain why this is one line with a `;`?

File: bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  /** Test peers. */
+  val testPeers: Set[Peer] = Set(
+    new Peer("foo", 123, Site.GN), 
+    new Peer("bar", 456, null)
+  )
+
+  /** Construct a KeyChain with a copy of the resource at the provided path. */
+  def keychain(resourcePath: String): Action[KeyChain] = {
+    val path = Files.createTempFile("KeyChainSpec", ".ser")
+    val file = path.toFile <| (_.deleteOnExit())
```
Comment (tpolecat) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: You boys are picky. Better?

Comment (cquiroz) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: 👍

Comment (arturog8m) [inline bundle/edu.gemini.util.security/src/test/scala/edu/gemini/util/security/auth/keychain/KeyChainSpec.scala]: Good. I can understand this version ;)!

**Pull request merged**